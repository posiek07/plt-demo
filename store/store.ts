import { Store } from 'pullstate';

export const MyBasket = new Store({
  basket: [],
  sum: 0,
});

export const deleteFromBasket = (productId) => {
  MyBasket.update((s) => {
    const newBasket = s;
    const removedItemIndex = newBasket.basket.findIndex((item) => item.id === productId);
    newBasket.basket.splice(removedItemIndex, 1);
    newBasket.sum =
      newBasket.basket.length > 0 ? newBasket.basket.reduce((sum, { price }) => sum + price, 0) : 0;
    return newBasket;
  });
};

export const addToBasket = (product) => {
  MyBasket.update((s) => {
    const newBasket = s;
    newBasket.basket.push(product);
    newBasket.sum =
      newBasket.basket.length > 0 ? newBasket.basket.reduce((sum, { price }) => sum + price, 0) : 0;
    return newBasket;
  });
};
