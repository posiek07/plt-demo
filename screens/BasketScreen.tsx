import * as React from 'react';
import { StyleSheet, SafeAreaView, Text, View } from 'react-native';
import EditItem from '../components/EditItem';
import { addToBasket, deleteFromBasket } from '../store/store';

import { MyBasket } from '../store/store';

export default function BasketScreen() {
  const store = MyBasket.useState();
  const sum = store.sum
  const basketItems = store.basket

  const basketItemsSet = new Set(basketItems);

  const basketItemsSetArray = Array.from(basketItemsSet);

  return (
    <View style={styles.container}>
      <View style={{ flex: 1, width: '100%' }}>
        {!!basketItemsSetArray &&
          basketItemsSetArray.map((product, index) => {
            const numberOfItems = basketItems.filter((obj) => obj.id === product.id).length;
            return (
              <View key={index} style={styles.productContainer}>
                <Text style={styles.title}>{product.name}</Text>
                <EditItem product={product} numberOfItems={numberOfItems} addToBasket={addToBasket}
                  deleteFromBasket={deleteFromBasket}
                />
              </View>
            );
          })}
      </View>
      <Text style={styles.sum}>
        SUM: {sum.toFixed(2)}
        {'\u00A3'}
      </Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
    padding: 10,
    width: '80%',
  },
  productContainer: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-evenly',
    alignItems: 'center',
    marginHorizontal: 20,
  },
  sum: {
    fontSize: 50,
    fontFamily: 'space-mono',
  },
});
