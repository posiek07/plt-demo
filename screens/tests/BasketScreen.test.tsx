import React from 'react';
import renderer from 'react-test-renderer';
import BasketScreen from '../BasketScreen';
import { MyBasket } from '../../store/store';



jest.mock("../../store/store.ts")

const mockState = {
    basket: [
        {
            id: 1,
            colour: 'Black',
            name: 'Black Sheet Strappy Textured Glitter Bodycon Dress',
            price: 10,
            img: 'http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024',
        },
        {
            id: 1,
            colour: 'Black',
            name: 'Black Sheet Strappy Textured Glitter Bodycon Dress',
            price: 10,
            img: 'http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024',
        },
        {
            colour: "Stone",
            id: 2,
            img: "https://cdn-img.prettylittlething.com/3/6/5/a/365a5d1dce6a2b77b564379b302c9d83afccf33b_cmd2051_1.jpg?imwidth=1024",
            name: "Stone Ribbed Strappy Cut Out Detail Bodycon Dress",
            price: 4,
        }
    ],
    sum: 20
}

MyBasket.useState = () => mockState

test('renders correctly', () => {
    const tree = renderer.create(<BasketScreen />).toJSON();
    expect(tree).toMatchSnapshot();
});
