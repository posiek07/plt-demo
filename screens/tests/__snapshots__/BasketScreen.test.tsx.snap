// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-evenly",
          "marginHorizontal": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "fontWeight": "bold",
            "padding": 10,
            "width": "80%",
          }
        }
      >
        Black Sheet Strappy Textured Glitter Bodycon Dress
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "padding": 10,
          }
        }
      >
        <Text
          onPress={[Function]}
          style={
            Object {
              "fontSize": 30,
            }
          }
        >
          +
        </Text>
        <Text
          style={
            Object {
              "fontSize": 30,
            }
          }
        >
          2
        </Text>
        <Text
          onPress={[Function]}
          style={
            Object {
              "fontSize": 30,
            }
          }
        >
          -
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-evenly",
          "marginHorizontal": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "fontWeight": "bold",
            "padding": 10,
            "width": "80%",
          }
        }
      >
        Black Sheet Strappy Textured Glitter Bodycon Dress
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "padding": 10,
          }
        }
      >
        <Text
          onPress={[Function]}
          style={
            Object {
              "fontSize": 30,
            }
          }
        >
          +
        </Text>
        <Text
          style={
            Object {
              "fontSize": 30,
            }
          }
        >
          2
        </Text>
        <Text
          onPress={[Function]}
          style={
            Object {
              "fontSize": 30,
            }
          }
        >
          -
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-evenly",
          "marginHorizontal": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "fontWeight": "bold",
            "padding": 10,
            "width": "80%",
          }
        }
      >
        Stone Ribbed Strappy Cut Out Detail Bodycon Dress
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "padding": 10,
          }
        }
      >
        <Text
          onPress={[Function]}
          style={
            Object {
              "fontSize": 30,
            }
          }
        >
          +
        </Text>
        <Text
          style={
            Object {
              "fontSize": 30,
            }
          }
        >
          1
        </Text>
        <Text
          onPress={[Function]}
          style={
            Object {
              "fontSize": 30,
            }
          }
        >
          -
        </Text>
      </View>
    </View>
  </View>
  <Text
    style={
      Object {
        "fontFamily": "space-mono",
        "fontSize": 50,
      }
    }
  >
    SUM: 
    20.00
    Â£
  </Text>
</View>
`;
